### PR TITLE
URL and case rules fix

### PR DESCRIPTION
**Pull request by pan93412**
This pull request introduces enhancements to the spelling and case rule detection logic, as well as improvements to documentation and database interfaces. The main focus is to support alternative case forms for terms, refine the exclusion logic for spellchecking, and clarify setup instructions.

**Documentation improvements:**

* Updated `README.md` to use correct repository URLs, add shell/env block formatting, and clarify installation/startup steps. Redundant setup instructions were removed for clarity. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L27-R40) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L49-R50) [[3]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L93-L108)
* Added documentation for "alternative case forms" in the case rule section, explaining that terms like `App` can also be written as `app`.

**Database and model changes:**

* The `CaseRule` interface and schema now support an optional `alternatives` field (array of strings) for alternative case forms.
* Updated the `ICaseDatabase` interface and its implementations (`CaseDatabase`, `MockCaseDatabase`) so `addRule` accepts an optional `alternatives` parameter. [[1]](diffhunk://#diff-6a2fbe395237fd44b75991a54b1ba95f969b8cf2fa5551cfa10aa6dd7a443969R83-R85) [[2]](diffhunk://#diff-e2f907ef1bfec78f067e422e265ecc1cddc29ea9164dda1c56db5a683f8d902fR140-R143) [[3]](diffhunk://#diff-c238e32e89a03d55107703b924a8396a347c830292e85c4c1d95fca7213d51c4L76-R77)

**Spellchecking logic improvements:**

* Refactored the `Determiner` class in `determiner.ts` to modularize and improve exclusion logic for URLs, paths, code blocks, and mentions. Added helper methods for building/merging exclusion ranges and for checking matches outside excluded ranges.
* Enhanced case rule checking to support alternative forms, ensuring that only incorrect case usages outside excluded ranges are flagged.

I also resolved the case-checking issues evident in the CI. There should be no spelling or casing problems in this codebase now.

**Minor dependency and import updates:**

* Adjusted imports in `determiner.ts` to include the new `CaseRule` type.

Fixed #26
Fixed https://discord.com/channels/1259032762422136902/1259032897373999144/1433821827364425758

## Summary by CodeRabbit

* **New Features**
  * Added support for alternative terms/casings when defining case rules.
  * Improved handling of special content areas—rules now exclude URLs, code blocks, inline code, and Markdown links from checks.

* **Documentation**
  * Updated setup instructions with corrected repository references and better formatting.

* **Tests**
  * Added comprehensive test coverage for edge cases including URLs, code blocks, and Markdown scenarios.

### CODE REVIEW

File: determiner.ts [hunk 1]
```diff
@@ -28,188 +32,249 @@ export class Determiner {
...
+		while ((match = pathRegex.exec(content)) !== null) {
 			const start = match.index;
 			const end = start + match[0].length;
-			// 跳過已被雙反引號或三反引號匹配的範圍
-			if (!isOverlapping(start, end)) {
-				excludedRanges.push({ start, end });
+
+			// 檢查路徑前是否有 <scheme>://（例如 https://github.com/user/repo）
+			const beforePath = content.slice(Math.max(0, start - BACKTRACK_LENGTH), start);
+			if (this.#URL_REGEX.test(beforePath)) {
```
[the remark this answers is not in the document]
Comment (pan93412) [reply determiner.ts:98]: 應該是還好

File: determiner.ts [hunk 2]
```diff
@@ -16,6 +16,10 @@ export class Determiner {
 	#spellingDatabase: ISpellingDatabase;
 	#caseDatabase: ICaseDatabase;
 
+	// https://datatracker.ietf.org/doc/html/rfc1738#section-2.1
+	// https://datatracker.ietf.org/doc/html/rfc1738#section-3.1
+	#URL_REGEX = /[a-zA-Z0-9+.-]*:\/\/[^\s]+/g;
```
[the remark this answers is not in the document]
Comment (pan93412) [reply determiner.ts]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/Edit-Mr/WinstonBot/pull/29#discussion_r2481782431)

File: models.ts [hunk 3]
```diff
@@ -46,13 +46,27 @@ export const SpellingRuleModel = mongoose.model<SpellingRule>("SpellingRule", sp
...
+		required: false,
+		validate: {
+			validator: function (v: unknown) {
+				if (v === undefined || v === null) return true;
+				if (Array.isArray(v)) return v.every(item => typeof item === "string");
+				return false;
+			},
+			message: "alternatives must be an array of strings, null, or undefined"
+		}
 	}
```
[the remark this answers is not in the document]
Comment (pan93412) [reply models.ts:70]: that's fine

**Review by elvisdragonmao — APPROVED** [review]
